File: 224. Basic Calculator.py
class Solution:
    def recursion(self, cur, s):
        plus, op = 0, 1
        
        while cur[0]<len(s):
            store = ''
            
            if s[cur[0]]=='(':
                cur[0]+=1
                plus += (op*self.recursion(cur, s))
            
            elif s[cur[0]]==')':
                return plus
            
            elif s[cur[0]]=='+' or s[cur[0]]=='-':
                op = 44-ord(s[cur[0]])
            
            while cur[0]<len(s) and s[cur[0]].isdigit():
                store, cur[0] = store+s[cur[0]], cur[0]+1
            
            if store:    
                plus, op, cur[0] = plus+(op*int(store)), 1, cur[0]-1

            cur[0] += 1
        
        return plus
    
    def calculate(self, s: str) -> int:
        return self.recursion([0], s)

File: 224. Basic Calculator.js
/**
 * @param {string} s
 * @return {number}
 */
var recursion = (s, cur)=>{
    let plus = 0, op = 1;

    while (cur[0]<s.length){
        let store = '';
        
        if (s[cur[0]] === '('){
            cur[0] += 1;
            plus += (op*recursion(s, cur));
        }else if (s[cur[0]] === ')'){
            return plus
        }else if (s[cur[0]]=='+'||s[cur[0]]=='-'){
            op = 44-s.charCodeAt(cur[0]);
        }
        
        while ( (cur[0]<s.length) && (!isNaN(parseInt(s[cur[0]]))) ){
            [store, cur[0]] = [store+s[cur[0]], cur[0]+1];
        }
        if (store){
            [plus, cur[0]] = [plus+op*parseInt(store, 10), cur[0]-1];
        }
        
        cur[0] += 1;
    }

    return plus;
}

var calculate = function(s) {
    return recursion(s, [0]);
};